src/Conversion/ONNXToKrnl/Tensor/ReverseSequence.cpp

```diff
@@ -87,41 +84,36 @@ struct ONNXReverseSequenceOpLowering : public ConversionPattern {
     */
 
     // Define loops and iteration trip counts (equivalent to size of output)
-    BuildKrnlLoop outputLoops(rewriter, loc, outputRank);
-    outputLoops.createDefineOp();
-    outputLoops.pushAllBounds(shapeHelper.dimsForOutput(0));
-    outputLoops.createIterateOp();
-
-    // Insert code inside the loop.
-    rewriter.setInsertionPointToStart(outputLoops.getIterateBlock());
-    IndexExprScope innerLoopScope(&rewriter, &outerScope);
-
-    LiteralIndexExpr one(1);
-
-    // LiteralIndexExpr batch(axisLit);
-    // SymbolIndexExpr axisDim(shapeHelper.dataDims[axisLit]);
-
-    // compute the loop indices for the output
-    SmallVector<IndexExpr, 4> outputAccessFct;
-    getIndexExprList<DimIndexExpr>(
-        outputLoops.getAllInductionVar(), outputAccessFct);
-
-    // Compute access function for indices[jjs].
-    SmallVector<IndexExpr, 4> inputAccessFct;
-    getIndexExprList<DimIndexExpr>(
-        outputLoops.getAllInductionVar(), inputAccessFct);
-    Value lensVal = createKrnl.loadIE(
-        operandAdaptor.sequence_lens(), inputAccessFct[batchAxis]);
-    IndexExpr lens = NonAffineIndexExpr(lensVal);
-    IndexExpr timeDim = inputAccessFct[timeAxis];
-    IndexExpr cond = timeDim < lens;
-    IndexExpr inputIndex =
-        IndexExpr::select(cond, lens - timeDim - one, timeDim);
-    inputAccessFct[timeAxis] = inputIndex;
-    Value inputVal = createKrnl.loadIE(operandAdaptor.input(), inputAccessFct);
-
-    // Save data into output
-    createKrnl.storeIE(inputVal, alloc, outputAccessFct);
+    KrnlBuilder createKrnl(rewriter, loc);
+    ValueRange loopDef = createKrnl.defineLoops(outputRank);
+    SmallVector<IndexExpr, 4> lbs(outputRank, LiteralIndexExpr(0));
+    createKrnl.iterateIE(loopDef, loopDef, lbs, shapeHelper.dimsForOutput(),
+        [&](KrnlBuilder &createKrnl, ValueRange loopInd) {
+          IndexExprScope innerLoopScope(&rewriter, shapeHelper.scope);
+          LiteralIndexExpr one(1);
```

Comment:
Loop invariant, possible to put outside the loop?

Reply:
https://github.com/onnx/onnx-mlir/pull/1305